### PR TITLE
facets: tree classification facet for experiments

### PR DESCRIPTION
**Pull request by miguelgrc**
Ref: #1072

### CODE REVIEW

File: ui/src/experiments/containers/SearchPageContainer.jsx [hunk 1]
```diff
@@ -19,7 +23,31 @@ function renderExperimentItem(result) {
...
+  const renderAggregations = useCallback(
+    () => (
+      <LoadingOrChildren loading={loadingAggregations}>
+        <AggregationFiltersContainer namespace={EXPERIMENTS_NS} />
+      </LoadingOrChildren>
+    ),
+    [loadingAggregations]
+  );
+
+  useEffect(
```
Comment (harunurhan) [inline ui/src/experiments/containers/SearchPageContainer.jsx:41]: unnecessary

File: ui/src/common/components/TreeAggregation.jsx [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+      .toArray();
+    onChange(selections);
+  }
+
+  getCheckedKeys(buckets) {
+    const { selectionMap } = this.state;
+    const checkedKeys = buckets
+      .map(bucket => bucket.get('key'))
+      .filter(bucketKey => selectionMap.get(bucketKey));
+    return checkedKeys.toJS();
```
Comment (harunurhan) [inline ui/src/common/components/TreeAggregation.jsx:68]: `toArray`

File: ui/src/common/components/TreeAggregation.jsx [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+import { Col, Row, Tree } from 'antd';
+
+import UnclickableTag from './UnclickableTag';
+import AggregationBox from './AggregationBox';
+import { forceArray } from '../utils';
+import FormattedNumber from './FormattedNumber';
+
+export const BUCKET_NAME_SPLITTER = '_';
+
+class TreeAggregation extends Component {
```
Comment (harunurhan) [inline ui/src/common/components/TreeAggregation.jsx:13]: we don't write new components this way anymore, instead prefer function components with hooks.

Comment (harunurhan) [reply ui/src/common/components/TreeAggregation.jsx:13]: you can start writing the tests, then refactor which should make it easier.

File: ui/src/common/components/TreeAggregation.jsx [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+      docCount,
+      splitTreeBy
+    );
+    return subTree;
+  }
+
+  render() {
+    const { name, buckets, splitTreeBy } = this.props;
+    let tree = [];
+    for (let i = 0; i < buckets.size; i += 1) {
```
Comment (harunurhan) [inline ui/src/common/components/TreeAggregation.jsx:117]: `render` called many times so in general doing a loop like this, could lead performance problems. and that should be memoized

File: ui/src/search/config.js [hunk 5]
```diff
@@ -79,7 +79,7 @@ const searchConfig = {
   },
   [EXPERIMENTS_NS]: {
     persistedQueryParamsDuringNewSearch: defaultPersistedQueryParamsDuringNewSearch,
-    onQueryChange: onAggregationlessCollectionQueryChange,
+    onQueryChange: onEventsQuerychange,
```
Comment (harunurhan) [inline ui/src/search/config.js:82]: maybe rename the `onEventsQuerychange` to something that fits to `seminars`, `conferences`, `experiments`

File: ui/src/experiments/containers/SearchPageContainer.jsx [hunk 6]
```diff
@@ -53,11 +94,25 @@ function ExperimentSearchPage({ loading }) {
...
   loading: state.search.getIn(['namespaces', EXPERIMENTS_NS, 'loading']),
+  loadingAggregations: state.search.getIn([
+    'namespaces',
+    EXPERIMENTS_NS,
+    'loadingAggregations',
+  ]),
+});
+
+const dispatchToProps = dispatch => ({
+  onBaseQueryChange(namespace, baseQuery) {
```
Comment (harunurhan) [inline ui/src/experiments/containers/SearchPageContainer.jsx:113]: also this., baseQueries only change for embedded searches.

File: ui/src/common/components/TreeAggregation.jsx (hunk 4 above)
Comment (harunurhan) [reply ui/src/common/components/TreeAggregation.jsx:117]: also this logic should be part of `buildTree` function

Comment (miguelgrc) [reply ui/src/common/components/TreeAggregation.jsx]: In which situations can that happen? The page is re-rendered when the filter changes and in that case we need to recreate the tree so it seems okay like this.

File: ui/src/search/config.js (hunk 5 above)
Comment (miguelgrc) [reply ui/src/search/config.js:82]: Maybe `onCollectionQueryChange`?

File: ui/src/common/components/TreeAggregation.jsx (hunk 4 above)
Comment (harunurhan) [reply ui/src/common/components/TreeAggregation.jsx]: you need to rebuild the tree only when `buckets` change, but this component's render method can be called for many different reasons, such other props change, internal state change, parent rerender etc.

File: ui/package.json [hunk 7]
```diff
@@ -36,7 +36,7 @@
     "craco-antd": "1.11.0",
     "formik": "^2.0.0",
     "history": "^4.7.2",
-    "immutable": "^3.8.2",
+    "immutable": "4.0.0-rc.10",
```
Comment (drjova) [inline ui/package.json:39]: rc?

Comment (harunurhan) [reply ui/package.json:39]: yeah, it's fine it has been `rc` for so long that I think it's stable :)

**Pull request merged**